Improve FunctionalCPD sampling by allowing vectorized functions

0149Sailesh opened the pull request:
### Your checklist for this pull request
Please review the [guidelines for contributing](CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [x] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [x] Check the commit's or even all commits' message styles matches our requested structure.

### Issue number(s) that this pull request fixes
- Fixes #1891 

### List of changes to the codebase in this pull request
- Introduced a new flag in FunctionalCPD that indicates if the supplied function allows vectorized input.
- Wrote 4 new tests to validate the working of the change. 

![image](https://github.com/user-attachments/assets/bf08b067-983b-40d5-82dd-7b76c9650060)

0149Sailesh committed:
feat(): Improve FunctionalCPD by allowing vectorization

0149Sailesh committed:
fix: Add tests with parent_sample=None to increase code coverage

ankurankan reviewed (changes requested):
@0149Sailesh Thanks for the PR. I think it mostly looks good and is showing significant runtime improvements. I would suggest simplifying the current implementation a bit though. I think we should just get rid of the parallel implementation and just keep two options: 1) vectorized, 2) iteratively sampling.

0149Sailesh commented:
Got it @ankurankan. Thanks for the suggestion. I'll make the required changes.

0149Sailesh committed:
fix: Address PR comments

0149Sailesh committed:
Merge branch 'dev' into feature/Improve_FunctionalCPD

0149Sailesh commented:
Hey @ankurankan 
I have addressed the PR comments. I did not remove the benchmark test as I feel it might help future devs. Let me know if you think otherwise or if the tests can be moved to a different folder where we can club together other benchmark tests.

0149Sailesh committed:
fix: Remove deprecated function

ankurankan commented on pgmpy/factors/hybrid/FunctionalCPD.py:
```diff
@@ -97,18 +98,40 @@ def sample(self, n_samples=100, parent_sample=None):
                 raise ValueError("Length of `parent_sample` must match `n_samples`.")
+            if self.vectorized:
+                dists = self.fn(parent_sample)
+                if isinstance(dists, list):
```
I don't think dists can ever be a list. Am I missing something here?

0149Sailesh replied:
Ahh yea. To be honest it is an unnecessary check. The only case I can think of is when the supplied function accidentally returns a list. This might cause pyro.sample to throw a TypeError.

ankurankan commented on pgmpy/factors/hybrid/FunctionalCPD.py:
```diff
@@ -97,18 +98,40 @@ def sample(self, n_samples=100, parent_sample=None):
-                )
+            if self.vectorized:
+                distribution = self.fn(None)
+                if isinstance(distribution, list):
```
Same here.

ankurankan commented on pgmpy/tests/test_factors/test_hybrid/test_Functional_CPD.py:
```diff
@@ -103,3 +104,117 @@ def test_different_distributions(self):
+        self.assertTrue(np.isfinite(samples).all())
+
+    # Benchmark test to see performance improvement between vectorized vs iterative. Use -s option to run this test to be able to see prints.
+    def test_sampling_time_vectorized_vs_iterative(self):
```
You can add benchmarks to pgmpy benchmarks (https://github.com/pgmpy/pgmpy-benchmarks), but it should not be in tests.

0149Sailesh replied:
Sure! I'll move it there

0149Sailesh committed:
fix(): Bye bye benchmark

0149Sailesh committed:
Merge branch 'dev' into feature/Improve_FunctionalCPD

ankurankan reviewed (approved):
Looks good to me now. Thanks for the PR @0149Sailesh
